ztag/transform.py

```diff
@@ -194,3 +196,175 @@ def transform(self, obj):
             out.transformed['domain'] = domain
         out.transformed['timestamp'] = obj['timestamp']
         return out
+
+
+class ZGrab2Transform(ZMapTransform):
+    """
+    This is registered in zschema as "zgrab2".
+    The generic format is
+    {
+        "ip": IPv4Address(),
+        "domain": String(),
+        "data": {
+            [scan-id]: {
+                "status": Enum(values=["success",
+                  "connection-refused",
+                  "connection-timeout",
+                  "connection-closed",
+                  "io-timeout",
+                  "protocol-error",
+                  "application-error",
+                  "unknown-error"
+                ]),
+                "protocol": [scan-protocol-name],
+                "timestamp": DateTime(),
+                "result": { [protocol-specific] },
+                "error": String(),
+            }
+        }
+    }
+    There are some standardized fields in result:
+        "tls" is a standard TLS structure (see https.make_tls_obj)
+        ...
+    """
+
+    # TODO -- pull from zgrab2_schemas.zgrab2.zgrab2.STATUS_VALUES?
+    STATUSES = {
+        "success",
+        "connection-refused",
+        "connection-timeout",
+        "connection-closed",
+        "io-timeout",
+        "protocol-error",
+        "application-error",
+        "unknown-error",
+    }
+
+    def __init__(self, *args, **kwargs):
+        """
+        Same as old ZGrabTransform.
+        :param args: Nothing new (forwarded to ZMapTransform)
+        :param kwargs: strip_domain_prefix -- optional domain prefix to strip
+                       (as with ZGrabTransform)
+        """
+        self.strip_domain_prefix = kwargs.get('strip_domain_prefix', '')
+        super(ZGrab2Transform, self).__init__(*args, **kwargs)
+
+    def invalid_result(self, obj, fmt=None, *args):
+        """
+        Throw an error indicating that this is not a valid ZGrab2 scan result.
+        :param obj: The noncompliant input JSON object
+        :param fmt: optional format string for the error message
+        :param args: optional format string arguments for the error message
+        :raises Exception:
+        """
+        if fmt is not None:
+            suffix = ": %s" % (fmt % args)
+        else:
+            suffix = ""
+        raise Exception("Not a valid ZGrab2 result" + suffix)
```

Comment:
My thinking was, if it has gotten this far, we are expecting a ZGrab2 result -- so malformed data here would be exceptional.

Having said that, the current practice seems to be for `IgnoreObject` to be raised in these circumstances -- so if we want to keep going that route, we can switch it...but my thinking was that we would want to catch these rather than silently failing.

Maybe we need something in between `IgnoreObject` and uncaught exception, like `ObjectFormatException`, which wouldn't cause a crash, but could be prominently logged?